proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+>a type variable that is already in scope does it have the effect of occluding
+>the outer name. Convenience and composability says yes, security says no.
+
+#### Interface {#interface}
+
+An interface type is a collection of type signatures, as an unordered set:
+
+```
+TypeInterface => '{' Signature ..  Signature '}'
+```
+
+```
+W-TypeInterface ::= '{' W-Signature .. W-Signature '}'
+```
+
+There are two forms of Signature; one denotes the type of a field: i.e., it’s an
+association of a type to a name:
+
+```
+Signature => Name ':' Type
+```
+
+```
+W-Signature ::= '(' W-Identifier W-Type ')'
+```
+
+The second form of TypeSignature denotes a nominal type, either implicitly:
+
+```
+Signature => 'type' NominalType
+```
+
+```
+W-Signature ::= '(' 'type' W-NominalType ')'
+```
+
+This form is used to identify so-called type imports. It is also used to
+identify nominal types; in which case it may be associated with an algebraic
+type combination:
+
+```
+Signature => 'type' NominalType '=' Type
+Signature => 'all' TypeVar .. TypeVar '.' NominalType '=' Type
+```
+
+```
+W-Signature ::= '(' 'type' W-NominalType W-Type ')'
+W-Signature ::= '(' 'type' '(' 'all' '(' TypeVar .. TypeVar ')' W-NominalType W-Type ')' ')'
+```
+
+For example, the standard `Option/1` type can be defined as the _Signature_:
+
+```
+( type (all (t) <Option t> [cases none (some t)]))
+```
+
+#### Tuple Type {#tuple-type}
+
+A TupleType is a linear sequence of types. It denotes the cross product of its
+component types:
+
+```
+TupleType => '(' Type ..  Type ')'
+```
+
+```
+W-TupleType ::= '(' W-Type .. W-Type ')'
+```
+
+There are two special cases of _TupleType_ that warrant further discussion: the
+unary tuple and the zero-ary tuple.
+
+A unary tuple type, such as
+
+```
+(integer)
+```
+
+is _not_ equivalent to its element type.
+
+The zero tuple type:
+
+```
+()
+```
+is used to designate void.
+
+>NOTE: Due to the ubiquity of tuples, the W-Expression form of a tuple type has
+>a special status: it is not marked with a special keyword as the first element
+>of its written form.
+
+#### Variant Type {#type-union}
+
+
+A variant type -- sometimes called a discriminated union -- is a collection of
+alternate forms of type; each of which is labeled by a discriminator.
+
+```
+VariantType => Name Type '|' .. '|' Name Type
+```
+
+If the Type is the empty tuple then it may be omitted.
+
+```
+W-VariantType ::= '[' 'cases' W-Variant .. W-Variant ']'
+W-Variant ::= W-Identifier | '(' W-Identifier W-Type ')'
+```
+
+<div class="example">
+The legal values for standard Boolean type may be defined as a VariantType:
+
+```
+[cases True False]
+```
+</div>
+
+If the Type is a TypInterface then the variant counts as the description of a
+record type; if the Type is a TupleType then the variant counts as a form of
+labeled tuple.
+
+#### Nominal Type {#nominal-type}
+
+A nominal type expression is used to identify types by name:
```

Comment:
It's not clear what you mean by "nominal". Is this just a type alias, or does it introduce a fresh type? "Nominal typing" refers to the latter, but the description kinda suggests the former. Especially since true nominal types would not easily allow self-contained interface descriptions.